File: agent-world-lab/backend/python/app/infrastructure/services/hitl_policy_service.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from app.application.interfaces import IHitlPolicyService
from app.application.models import CurrentUser
from app.domain.models import RiskLevel, ToolDefinition
from app.settings import HitlSettings


class HitlPolicyService(IHitlPolicyService):
    def __init__(self, options: HitlSettings) -> None:
        self._options = options

    def requires_approval(self, user: CurrentUser, tool: ToolDefinition, environment: str | None) -> bool:
        mode = self._options.mode.lower()
        if mode == "none":
            return False
        if mode == "always":
            return True
        if mode == "risky-tools-only":
            return tool.risk_level == RiskLevel.HIGH
        if mode == "role-based":
            return not user.is_in_role("Admin") and tool.risk_level != RiskLevel.LOW
        if mode == "environment-based":
            return (environment or "").lower() == "prod"
        return False

    def compute_approval_expiry(self) -> datetime:
        return datetime.now(timezone.utc) + timedelta(minutes=self._options.approval_expiry_minutes)

File: agent-world-lab/backend/src/AgentWorldLab.Infrastructure/Services/HitlPolicyService.cs
using AgentWorldLab.Application.Abstractions;
using AgentWorldLab.Application.Models;
using AgentWorldLab.Domain.Models;
using Microsoft.Extensions.Options;

namespace AgentWorldLab.Infrastructure.Services;

public sealed class HitlPolicyService : IHitlPolicyService
{
    private readonly HitlOptions _options;

    public HitlPolicyService(IOptions<HitlOptions> options)
    {
        _options = options.Value;
    }

    public bool RequiresApproval(CurrentUser user, ToolDefinition tool, string environment)
    {
        var mode = _options.Mode.Trim().ToLowerInvariant();
        return mode switch
        {
            "none" => false,
            "always" => true,
            "risky-tools-only" => tool.RiskLevel == RiskLevel.High,
            "role-based" => !user.IsInRole("Admin") && tool.RiskLevel != RiskLevel.Low,
            "environment-based" => string.Equals(environment, "prod", StringComparison.OrdinalIgnoreCase),
            _ => tool.RiskLevel == RiskLevel.High
        };
    }

    public DateTimeOffset ComputeApprovalExpiry()
    {
        return DateTimeOffset.UtcNow.AddMinutes(Math.Max(_options.ApprovalExpiryMinutes, 1));
    }
}
